### ISSUE TITLE
Could not load genomicsdb native library

### ISSUE THREAD

**Issue by komalsrathi** [bcbio/bcbio-nextgen#2456]
Hi,

Another issue for the RNA-seq variant calling regarding loading genomicsdb, logs are attached here:

[logs.txt](https://github.com/bcbio/bcbio-nextgen/files/2199087/logs.txt)

Please let me know if you need more information.

**Comment by roryk** [comment]
Hi Komal,

Hm, I think the clue is here in the log:

```
Runtime.totalMemory()=460259328
Exception in thread "main" java.lang.ExceptionInInitializerError
        at org.broadinstitute.hellbender.tools.genomicsdb.GenomicsDBImport.overwriteOrCreateWorkspace(GenomicsDBImport.java:660)
        at org.broadinstitute.hellbender.tools.genomicsdb.GenomicsDBImport.onTraversalStart(GenomicsDBImport.java:461)
        at org.broadinstitute.hellbender.engine.GATKTool.doWork(GATKTool.java:982)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:135)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:180)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:199)
        at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
        at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
        at org.broadinstitute.hellbender.Main.main(Main.java:289)
Caused by: com.intel.genomicsdb.exception.GenomicsDBException: Could not load genomicsdb native library
        at com.intel.genomicsdb.importer.GenomicsDBImporter.<clinit>(GenomicsDBImporter.java:65)
        ... 9 more
Caused by: java.lang.UnsatisfiedLinkError: no tiledbgenomicsdb in java.library.path
        at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
        at java.lang.Runtime.loadLibrary0(Runtime.java:870)
        at java.lang.System.loadLibrary(System.java:1122)
        at com.intel.genomicsdb.GenomicsDBLibLoader.loadLibrary(GenomicsDBLibLoader.java:43)
        at com.intel.genomicsdb.importer.GenomicsDBImporter.<clinit>(GenomicsDBImporter.java:62)
```

Do you happen to have `LD_LIBRARY_PATH` set?

**Comment by komalsrathi** [comment]
Yes.

```
[rathik@reslnrab01 ~]$ echo $LD_LIBRARY_PATH 
:/cm/shared/apps/intel-tbb-oss/intel64/current/lib/gcc4.4:/usr/lib64:/cm/shared/apps/intel-tbb-oss/intel64/2017_20170807oss/lib/intel64/gcc4.4

```

**Comment by roryk** [comment]
Can you unset LD_LIBRARY_PATH before running? It looks like GATK is looking for a shared library and can't find it, sometimes having LD_LIBRARY_PATH set can cause these kinds of issues.

**Comment by komalsrathi** [comment]
I tried this and it didn't work. I am clueless as to what is going on. If you aren't able to debug this then you can close it.

**Comment by roryk** [comment]
Hi Komal,

Guh, sorry for the problems. If you go into the work/log directory and do this:

```
grep GenomicsDB bcbio-nextgen-commands.log
```

it should show the failing command. For example on our tests:

```
unset JAVA_HOME && export PATH=/home/rdk4/local/share/bcbio/anaconda/bin:$PATH && gatk --java-options '-Xms454m -Xmx3181m -XX:+UseSerialGC -Djava.io.tmpdir=/n/app/bcbio/dev/rory-dev/bcbio-nextgen/tests/test_automated_output/joint/gatk-haplotype-joint/TestBatch1/chrM/bcbiotx/tmp312_s1' GenomicsDBImport --reader-threads 1 --genomicsdb-workspace-path TestBatch1-chrM_0_16571_genomicsdb -L chrM:1-16571 --variant /n/app/bcbio/dev/rory-dev/bcbio-nextgen/tests/test_automated_output/variation/rnaseq/gatk-haplotype/Test1-gatk-haplotype.vcf.gz --variant /n/app/bcbio/dev/rory-dev/bcbio-nextgen/tests/test_automated_output/variation/rnaseq/gatk-haplotype/Test2-gatk-haplotype.vcf.gz
```

Could you post what that is for you? And if you run that command by itself, does it fail?

**Comment by komalsrathi** [comment]
Here:

```
[2018-07-16T18:08Z] unset JAVA_HOME && export PATH=/mnt/isilon/cbmi/variome/bin/bcbio-nextgen/bcbio/anaconda/bin:$PATH && gatk --java-options '-Xms454m -Xmx3181m -XX:+UseSerialGC -Djava.io.tmpdir=/mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/joint/gatk-haplotype-joint/CDL-634-13P/1/bcbiotx/tmp5vG0IL' GenomicsDBImport --reader-threads 1 --genomicsdb-workspace-path CDL-634-13P-1_0_249250621_genomicsdb -L 1:1-249250621 --variant /mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/variation/rnaseq/gatk-haplotype/Sample_1__CDL-634-13P-gatk-haplotype-annotated-rnaedit-annotated-gemini.vcf.gz
```

**Comment by roryk** [comment]
Thanks! What happens if you run that command outside of bcbio?

**Comment by komalsrathi** [comment]
It crashes with the same error:

```
[rathik@reslnrefo01 log]$ unset JAVA_HOME && export PATH=/mnt/isilon/cbmi/variome/bin/bcbio-nextgen/bcbio/anaconda/bin:$PATH && gatk --java-options '-Xms1g -Xmx99g -XX:+UseSerialGC -Djava.io.tmpdir=/mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/joint/gatk-haplotype-joint/CDL-634-13P/1/bcbiotx/tmpHA8ElG' GenomicsDBImport --reader-threads 1 --genomicsdb-workspace-path CDL-634-13P-1_0_249250621_genomicsdb -L 1:1-249250621 --variant /mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/variation/rnaseq/gatk-haplotype/Sample_1__CDL-634-13P-gatk-haplotype-annotated-rnaedit-annotated-gemini.vcf.gz
Using GATK jar /mnt/isilon/cbmi/variome/bin/bcbio-nextgen/bcbio/anaconda/share/gatk4-4.0.6.0-0/gatk-package-4.0.6.0-local.jar
Running:
    java -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=true -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=2 -Xms1g -Xmx99g -XX:+UseSerialGC -Djava.io.tmpdir=/mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/joint/gatk-haplotype-joint/CDL-634-13P/1/bcbiotx/tmpHA8ElG -jar /mnt/isilon/cbmi/variome/bin/bcbio-nextgen/bcbio/anaconda/share/gatk4-4.0.6.0-0/gatk-package-4.0.6.0-local.jar GenomicsDBImport --reader-threads 1 --genomicsdb-workspace-path CDL-634-13P-1_0_249250621_genomicsdb -L 1:1-249250621 --variant /mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/variation/rnaseq/gatk-haplotype/Sample_1__CDL-634-13P-gatk-haplotype-annotated-rnaedit-annotated-gemini.vcf.gz
Picked up _JAVA_OPTIONS: -Djava.io.tmpdir=/mnt/isilon/cbmi/variome/tmp/rathik
15:00:09.002 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/mnt/isilon/cbmi/variome/bin/bcbio-nextgen/bcbio/anaconda/share/gatk4-4.0.6.0-0/gatk-package-4.0.6.0-local.jar!/com/intel/gkl/native/libgkl_compression.so
15:00:09.373 INFO  GenomicsDBImport - ------------------------------------------------------------
15:00:09.373 INFO  GenomicsDBImport - The Genome Analysis Toolkit (GATK) v4.0.6.0
15:00:09.373 INFO  GenomicsDBImport - For support and documentation go to https://software.broadinstitute.org/gatk/
15:00:09.374 INFO  GenomicsDBImport - Executing as rathik@reslnrefo01.research.chop.edu on Linux v3.10.0-862.el7.x86_64 amd64
15:00:09.374 INFO  GenomicsDBImport - Java runtime: OpenJDK 64-Bit Server VM v1.8.0_144-b01
15:00:09.375 INFO  GenomicsDBImport - Start Date/Time: July 17, 2018 3:00:08 PM EDT
15:00:09.375 INFO  GenomicsDBImport - ------------------------------------------------------------
15:00:09.375 INFO  GenomicsDBImport - ------------------------------------------------------------
15:00:09.375 INFO  GenomicsDBImport - HTSJDK Version: 2.16.0
15:00:09.375 INFO  GenomicsDBImport - Picard Version: 2.18.7
15:00:09.375 INFO  GenomicsDBImport - HTSJDK Defaults.COMPRESSION_LEVEL : 2
15:00:09.375 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
15:00:09.375 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
15:00:09.375 INFO  GenomicsDBImport - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
15:00:09.376 INFO  GenomicsDBImport - Deflater: IntelDeflater
15:00:09.376 INFO  GenomicsDBImport - Inflater: IntelInflater
15:00:09.376 INFO  GenomicsDBImport - GCS max retries/reopens: 20
15:00:09.376 INFO  GenomicsDBImport - Using google-cloud-java patch 6d11bef1c81f885c26b2b56c8616b7a705171e4f from https://github.com/droazen/google-cloud-java/tree/dr_all_nio_fixes
15:00:09.376 INFO  GenomicsDBImport - Initializing engine
15:00:10.365 INFO  IntervalArgumentCollection - Processing 249250621 bp from intervals
15:00:10.370 INFO  GenomicsDBImport - Done initializing engine
Created workspace /mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/log/CDL-634-13P-1_0_249250621_genomicsdb
15:00:11.673 INFO  GenomicsDBImport - Vid Map JSON file will be written to CDL-634-13P-1_0_249250621_genomicsdb/vidmap.json
15:00:11.674 INFO  GenomicsDBImport - Callset Map JSON file will be written to CDL-634-13P-1_0_249250621_genomicsdb/callset.json
15:00:11.674 INFO  GenomicsDBImport - Complete VCF Header will be written to CDL-634-13P-1_0_249250621_genomicsdb/vcfheader.vcf
15:00:11.674 INFO  GenomicsDBImport - Importing to array - CDL-634-13P-1_0_249250621_genomicsdb/genomicsdb_array
15:00:11.674 INFO  ProgressMeter - Starting traversal
15:00:11.675 INFO  ProgressMeter -        Current Locus  Elapsed Minutes     Batches Processed   Batches/Minute
15:00:12.505 INFO  GenomicsDBImport - Importing batch 1 with 1 samples
Buffer resized from 22684bytes to 32523
Buffer resized from 32523bytes to 32671
Buffer resized from 32671bytes to 32769
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fc5ff88d359, pid=30328, tid=0x00007fc5ff737700
#
# JRE version: OpenJDK Runtime Environment (8.0_144-b01) (build 1.8.0_144-b01)
# Java VM: OpenJDK 64-Bit Server VM (25.144-b01 mixed mode linux-amd64 )
# Problematic frame:
# C  [libtiledbgenomicsdb7066406292930633618.so+0x155359]  BufferVariantCell::set_cell(void const*)+0x99
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /mnt/isilon/cbmi/variome/rathik/mendelian_rnaseq/bcbio_out/log/hs_err_pid30328.log
#
# If you would like to submit a bug report, please visit:
#   http://www.azulsystems.com/support/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
```

**Comment by roryk** [comment]
Hm, it looks like there have been some other issues in GenomicsDBImport: 
https://github.com/broadinstitute/gatk/issues/4994

**Comment by roryk** [comment]
The error running it alone is different than the one when it is run inside of bcbio, though.

**Comment by chapmanb** [comment]
Komal;
Thanks for all these detailed reports and apologies about the issues with GenomicsDB. I'm not totally positive what is going on but for the two issues:

1. For the core dump that you're seeing in #2453 and above (`BufferVariantCell::set_cell(void const*)+0x99`) I wonder if this is an issue with using too much memory. How much available memory is on your compute machines? In the example above you're asking for 99Gb of memory and in #2453 it's 53Gb of memory. Is it possible that other processes are running on these machines or other things restrict memory?

2. For the `no tiledbgenomicsdb in java.library.path` it seems to be running into problems finding appropriate shared libraries that are meant to be bundled with the jar file. What type of machine are you running this on? If we can provide more system information I think the best approach would be to raise an issue in the gatk repository to see if they can help with tracking down what is happening.

Thanks again for the help debugging this.

**Comment by komalsrathi** [comment]
Thanks. 

For 1, I actually asked one of our systems engineer to help me debug it as well and he said the box I was running on has like 700GB of free memory. I did try to use `-n 1` and `jvm_opts: ["-Xms2g", "-Xmx4g"]` so not sure how it could be a memory issue. 

For 2, not sure if this helps:

```
[rathik@reslnrab01 ~]$ lscpu

Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                96
On-line CPU(s) list:   0-95
Thread(s) per core:    2
Core(s) per socket:    12
Socket(s):             4
NUMA node(s):          4
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 5118 CPU @ 2.30GHz
Stepping:              4
CPU MHz:               2300.000
BogoMIPS:              4600.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              16896K
NUMA node0 CPU(s):     0,4,8,12,16,20,24,28,32,36,40,44,48,52,56,60,64,68,72,76,80,84,88,92
NUMA node1 CPU(s):     1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61,65,69,73,77,81,85,89,93
NUMA node2 CPU(s):     2,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,66,70,74,78,82,86,90,94
NUMA node3 CPU(s):     3,7,11,15,19,23,27,31,35,39,43,47,51,55,59,63,67,71,75,79,83,87,91,95
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 intel_ppin intel_pt mba tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local ibpb ibrs stibp dtherm ida arat pln pts pku ospke spec_ctrl intel_stibp
```

Also, how do I find `tiledbgenomicsdb`? I want to check where it is present on both the boxes so I can compare why is it missing.

**Comment by komalsrathi** [comment]
Hi Rory, 

I am closing this as I was able to circumvent this error by resetting my Java tmpdir. However, https://github.com/bcbio/bcbio-nextgen/issues/2453 still persists. But at least, now there is 1 less error to debug.

Thanks
Komal
